File: VB/MVVMExpenses/Common/DataModel/EntityFramework/DbUnitOfWorkFactory.vb
Imports System
Imports System.Linq
Imports System.Data
Imports System.Data.Entity
Imports System.Linq.Expressions
Imports System.Collections.Generic
Imports Common.Utils
Imports Common.DataModel
Imports DevExpress.Mvvm
Imports System.Collections
Imports System.ComponentModel
Imports DevExpress.Data.Linq
Imports DevExpress.Data.Linq.Helpers
Imports DevExpress.Data.Async.Helpers
Namespace Common.DataModel.EntityFramework
    Friend Class DbUnitOfWorkFactory(Of TUnitOfWork As IUnitOfWork)
        Implements IUnitOfWorkFactory(Of TUnitOfWork)
        Private _createUnitOfWork As Func(Of TUnitOfWork)
        Public Sub New(ByVal createUnitOfWork As Func(Of TUnitOfWork))
            Me._createUnitOfWork = createUnitOfWork
        End Sub
        Private Function CreateUnitOfWork() As TUnitOfWork Implements IUnitOfWorkFactory(Of TUnitOfWork).CreateUnitOfWork
            Return _createUnitOfWork()
        End Function
        Private Function CreateInstantFeedbackSource(Of TEntity As {Class, New}, TProjection As Class, TPrimaryKey)(ByVal getRepositoryFunc As Func(Of TUnitOfWork, IRepository(Of TEntity, TPrimaryKey)), ByVal projection As Func(Of IRepositoryQuery(Of TEntity), IQueryable(Of TProjection))) As IInstantFeedbackSource(Of TProjection) Implements IUnitOfWorkFactory(Of TUnitOfWork).CreateInstantFeedbackSource
            Dim threadSafeProperties = New TypeInfoProxied(TypeDescriptor.GetProperties(GetType(TProjection)), Nothing).UIDescriptors
            If projection Is Nothing Then
                projection = Function(x) TryCast(x, IQueryable(Of TProjection))
            End If
            Dim source = New EntityInstantFeedbackSource(Sub(e As GetQueryableEventArgs) e.QueryableSource = projection(getRepositoryFunc(_createUnitOfWork()))) With {.KeyExpression = getRepositoryFunc(_createUnitOfWork()).GetPrimaryKeyPropertyName()}
            Return New InstantFeedbackSource(Of TProjection)(source, threadSafeProperties)
        End Function
    End Class
End Namespace

File: CS/MVVMExpenses/Common/DataModel/EntityFramework/DbUnitOfWorkFactory.cs
using System;
using System.Linq;
using System.Data;
using System.Data.Entity;
using System.Linq.Expressions;
using System.Collections.Generic;
using MVVMExpenses.Common.Utils;
using MVVMExpenses.Common.DataModel;
using DevExpress.Mvvm;
using System.Collections;
using System.ComponentModel;
using DevExpress.Data.Linq;
using DevExpress.Data.Linq.Helpers;
using DevExpress.Data.Async.Helpers;

namespace MVVMExpenses.Common.DataModel.EntityFramework {
    class DbUnitOfWorkFactory<TUnitOfWork> : IUnitOfWorkFactory<TUnitOfWork> where TUnitOfWork : IUnitOfWork {
        Func<TUnitOfWork> createUnitOfWork;

        public DbUnitOfWorkFactory(Func<TUnitOfWork> createUnitOfWork) {
            this.createUnitOfWork = createUnitOfWork;
        }

        TUnitOfWork IUnitOfWorkFactory<TUnitOfWork>.CreateUnitOfWork() {
            return createUnitOfWork();
        }

        IInstantFeedbackSource<TProjection> IUnitOfWorkFactory<TUnitOfWork>.CreateInstantFeedbackSource<TEntity, TProjection, TPrimaryKey>(
            Func<TUnitOfWork, IRepository<TEntity, TPrimaryKey>> getRepositoryFunc,
            Func<IRepositoryQuery<TEntity>, IQueryable<TProjection>> projection) {
            var threadSafeProperties = new TypeInfoProxied(TypeDescriptor.GetProperties(typeof(TProjection)), null).UIDescriptors;
            if(projection == null) {
                projection = x => x as IQueryable<TProjection>;
            }
            var source = new EntityInstantFeedbackSource((GetQueryableEventArgs e) => e.QueryableSource = projection(getRepositoryFunc(createUnitOfWork()))) {
                KeyExpression = getRepositoryFunc(createUnitOfWork()).GetPrimaryKeyPropertyName(),
            };
            return new InstantFeedbackSource<TProjection>(source, threadSafeProperties);
        }
    }
}